### PR TITLE
feat(onebot): support time in forward messages

### PR DESCRIPTION
**Pull request by Seidko**
Implement the feature of changing the time of forwarding messages and forwarding messages through message ID.
image

### CODE REVIEW

File: adapters/onebot/src/bot/message.ts [hunk 1]
```diff
@@ -141,8 +153,10 @@ export class OneBotMessenger extends Messenger<BaseBot> {
         await this.flush()
         this.stack.shift()
         await this.forward()
+      } else if ('id' in attrs) {
+        this.stack[0].data = attrs.id.toString()
       } else {
-        Object.assign(this.stack[0].author, pick(attrs, ['userId', 'username', 'nickname', 'avatar']))
+        Object.assign(this.stack[0].data, pick(attrs, ['userId', 'user-id', 'username', 'nickname', 'time']))
```
Comment (MaikoTan) [inline adapters/onebot/src/bot/message.ts:159]: Why removed `avatar`?

File: adapters/onebot/src/bot/message.ts [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import { Messenger, pick, segment, Universal } from '@satorijs/satori'
 import { BaseBot } from './base'
 import { CQCode } from './cqcode'
+import { Author } from '../types'
 
 class State {
-  author: Partial<Universal.Author> = {}
+  data: Partial<Author> | string = {}
```
Comment (MaikoTan) [inline adapters/onebot/src/bot/message.ts:7]: I think you could maintain the original name `author` here. As `State` here are for saving miscellaneous pieces of information.

File: adapters/onebot/src/bot/message.ts [hunk 3]
```diff
@@ -43,17 +44,28 @@ export class OneBotMessenger extends Messenger<BaseBot> {
     }
 
     // flush
-    if (!this.children.length) return
-    const { type, author } = this.stack[0]
+    const { type, data } = this.stack[0]
+    if (!this.children.length && typeof data !== 'number') return
```
Comment (MaikoTan) [inline adapters/onebot/src/bot/message.ts:48]: Isn't the author can be only an `Author` or `string`? Where the hat that `number` from?

File: adapters/onebot/src/bot/message.ts [hunk 4]
```diff
@@ -43,17 +44,28 @@ export class OneBotMessenger extends Messenger<BaseBot> {
...
+          },
+        })
+      } else {
+        this.stack[1].children.push({
+          type: 'node',
+          data: {
+            id: data
+          },
+        })
+      }
```
Comment (MaikoTan) [inline adapters/onebot/src/bot/message.ts:67]: The simpler part should be put avant

**Review by MaikoTan — CHANGES_REQUESTED** [review]
Looks good, just several comments.

File: adapters/onebot/src/bot/message.ts (hunk 1 above)
Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:159]: because go-cqhttp didn't provide api to set avatar:
image
if you want set message avatar, just need to set `userId` .

Comment (MaikoTan) [reply adapters/onebot/src/bot/message.ts:159]: Ah ok.

File: adapters/onebot/src/bot/message.ts (hunk 2 above)
Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:7]: But if `author` is used as the `messageId`, it doesn't quite fit its meaning. Also, the name used by `gocq` is also `data`, so I have named it `data` as well.

Comment (MaikoTan) [reply adapters/onebot/src/bot/message.ts:7]: What do you mean by `used as the messageId`?
As my understanding, the `author` here is referring to the message element `author`, so everything you extracted from the element should be put in it. `data` is instead too broad here.

Comment (MaikoTan) [reply adapters/onebot/src/bot/message.ts:7]: In fact, you could just save the messageId in `author` just as you did for other parameters.

Comment (Seidko) [reply adapters/onebot/src/bot/message.ts]: well okey

File: adapters/onebot/src/bot/message.ts [hunk 5]
```diff
@@ -1,9 +1,10 @@
 import { Messenger, pick, segment, Universal } from '@satorijs/satori'
 import { BaseBot } from './base'
 import { CQCode } from './cqcode'
+import { Author } from '../types'
 
 class State {
-  author: Partial<Universal.Author> = {}
+  author: Partial<Author> | string = {}
```
Comment (MaikoTan) [inline adapters/onebot/src/bot/message.ts:7]: I reread the mechanic about this PR.
I think you could just make `author` as `Author`, which has a property `messageId`, assign the `author.messageId = attr.id.toString()`. then just `if (state.messageId) ... else ...`

You even no need to check the type of the `author` variable.

Comment (Seidko) [reply adapters/onebot/src/bot/message.ts]: lgtm

File: adapters/onebot/src/bot/message.ts [hunk 6]
```diff
@@ -43,17 +44,28 @@ export class OneBotMessenger extends Messenger<BaseBot> {
     }
 
     // flush
-    if (!this.children.length) return
     const { type, author } = this.stack[0]
+    if (!this.children.length && !author.messageId) return
```
Comment (MaikoTan) [inline adapters/onebot/src/bot/message.ts:48]: swap these lines, so if there is an early return happened, the js engine no need to create the variables and destroy them immediately.


```suggestion
    if (!this.children.length && !author.messageId) return
    const { type, author } = this.stack[0]
```

Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:48]: How can I get `author.messageId` if I do that?

Comment (MaikoTan) [reply adapters/onebot/src/bot/message.ts:48]: Oh, make sense.
It would be too wordy to write `this.stack[0].author`.

**Review by MaikoTan — APPROVED** [review]

**Comment by Seidko** [comment]
Okey, thanks for your help!

File: adapters/onebot/src/types.ts [hunk 7]
```diff
@@ -89,6 +89,11 @@ export interface UnidirectionalFriendInfo extends AccountInfo {
   source: string
 }
 
+export interface Author extends Universal.Author {
+  time?: string | number
+  messageId?: string
```
Comment (shigma) [inline adapters/onebot/src/types.ts:94]: What's the use of this `messageId`?

File: adapters/onebot/src/bot/message.ts [hunk 8]
```diff
@@ -141,8 +153,10 @@ export class OneBotMessenger extends Messenger<BaseBot> {
         await this.flush()
         this.stack.shift()
         await this.forward()
+      } else if ('id' in attrs) {
+        this.stack[0].author.messageId = attrs.id.toString()
       } else {
-        Object.assign(this.stack[0].author, pick(attrs, ['userId', 'username', 'nickname', 'avatar']))
+        Object.assign(this.stack[0].author, pick(attrs, ['userId', 'user-id', 'username', 'nickname', 'time']))
```
Comment (shigma) [inline adapters/onebot/src/bot/message.ts:159]: `'userId', 'user-id'`?

File: adapters/onebot/src/types.ts (hunk 7 above)
Comment (shigma) [reply adapters/onebot/src/types.ts:94]: If you are to send forward message with id, you should use `<message id>` rather than `<author message-id>`.

**Review by shigma — CHANGES_REQUESTED** [review]

File: adapters/onebot/src/bot/message.ts (hunk 8 above)
Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:159]: From document:
image
The `user-id` is used here. Of course, for compatibility, I have kept `userId`.

File: adapters/onebot/src/types.ts (hunk 7 above)
Comment (Seidko) [reply adapters/onebot/src/types.ts:94]: > What's the use of this `messageId`?

From go-cqhttp document:
image

> If you are to send forward message with id, you should use `<message id>` rather than `<author message-id>`.

Yes, that's what I did.

File: adapters/onebot/src/bot/message.ts (hunk 8 above)
Comment (shigma) [reply adapters/onebot/src/bot/message.ts:159]: `user-id` will be transformed into `userId`

Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:159]: But the issue now is that the user-id has not been transformed, do I need to handle this?

Comment (MaikoTan) [reply adapters/onebot/src/bot/message.ts:159]: You mean that there is only `user-id` in `attrs` instead of `userId`? If true, I think there is bugs in another place.

Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:159]: yes, I found that there is only `user-id` and no `userId`. It never been transformed.

Comment (shigma) [reply adapters/onebot/src/bot/message.ts:159]: Repro: param-case in the string form will be transformed, but JSX does not. Will be fixed on the @satorijs/element side.

Comment (shigma) [reply adapters/onebot/src/bot/message.ts:159]:
```suggestion
        Object.assign(this.stack[0].author, pick(attrs, ['userId', 'username', 'nickname', 'time']))
```

File: adapters/onebot/src/bot/message.ts [hunk 9]
```diff
@@ -43,17 +44,28 @@ export class OneBotMessenger extends Messenger<BaseBot> {
...
-          name: author.nickname || author.username || this.bot.nickname || this.bot.username,
-          uin: author.userId || this.bot.userId,
-          content: this.children as any,
-        },
-      })
+      if (author.messageId) {
+        this.stack[1].children.push({
+          type: 'node',
+          data: {
+            id: author.messageId
```
Comment (shigma) [inline adapters/onebot/src/bot/message.ts:54]:
```suggestion
            id: author.messageId,
```

File: adapters/onebot/src/bot/message.ts [hunk 10]
```diff
@@ -43,17 +44,28 @@ export class OneBotMessenger extends Messenger<BaseBot> {
...
+          },
+        })
+      } else {
+        this.stack[1].children.push({
+          type: 'node',
+          data: {
+            name: author.nickname || author.username || this.bot.nickname || this.bot.username,
+            uin: author.userId || author['user-id'] || this.bot.userId,
+            content: this.children as any,
+            time: `${(+author.time || Date.now()) / 1000}`
```
Comment (shigma) [inline adapters/onebot/src/bot/message.ts:64]:
```suggestion
            uin: author.userId || this.bot.userId,
            content: this.children as any,
            time: `${Math.floor((+author.time || Date.now()) / 1000)}`,
```

Not sure if go-cqhttp supports this.

**Review by shigma — CHANGES_REQUESTED** [review]

File: adapters/onebot/src/bot/message.ts (hunk 8 above)
Comment (Seidko) [reply adapters/onebot/src/bot/message.ts:159]: > Repro: param-case in the string form will be transformed, but JSX does not. Will be fixed on the @satorijs/element side.

Should I open an issue for this?

**Review by MaikoTan — APPROVED** [review]
